### PR TITLE
 Implement querying inboundgroupsessions by room_id

### PR DESCRIPTION
**Pull request by multisme**
History sharing: improve efficiency of building key bundle

Signed-off-by: multi [multiestunhappydev@gmail.com](mailto:multiestunhappydev@gmail.com)

Partially Implement https://github.com/matrix-org/matrix-rust-sdk/issues/5513

### CODE REVIEW

**Comment by multisme** [comment]
@richvdh 
Hi,
Am I on the right path regarding the implementation of that functionality with IndexedDB?

File: crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs [hunk 1]
```diff
@@ -967,6 +967,31 @@ impl_crypto_store! {
         }
     }
 
+    async fn get_inbound_group_session_by_room_id(
```
Comment (richvdh) [inline crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs]:
```suggestion
    async fn get_inbound_group_sessions_by_room_id(
```

Comment (richvdh) [reply crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs]: Also, this belongs *after* `get_inbound_group_sessions`.

**Review by richvdh — COMMENTED** [review]
Looks plausible to me. It will need defining in the trait, and some tests, obviously.

File: crates/matrix-sdk-crypto/src/store/traits.rs [hunk 2]
```diff
@@ -121,6 +121,15 @@ pub trait CryptoStore: AsyncTraitDeps {
...
     async fn get_inbound_group_sessions(&self) -> Result<Vec<InboundGroupSession>, Self::Error>;
 
+    /// Get all the inbound group sessions linked to a room
+    ///
+    /// # Arguments
+    /// * `room_id` - the room id of the room that the sessions are linked to.
+    async fn get_inbound_group_sessions_by_room_id(
+        &self,
+        room_id: &RoomId,
+    ) -> Result<Vec<InboundGroupSession>, Self::Error>;
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/traits.rs]: suggest slotting this in between `inbound_group_session_counts` and `get_inbound_group_sessions_for_device_batch`.

Likewise in `MemoryStore` and `SqliteCryptoStore`.

File: crates/matrix-sdk-crypto/src/store/traits.rs [hunk 3]
```diff
@@ -121,6 +121,15 @@ pub trait CryptoStore: AsyncTraitDeps {
     /// Get all the inbound group sessions we have stored.
     async fn get_inbound_group_sessions(&self) -> Result<Vec<InboundGroupSession>, Self::Error>;
 
+    /// Get all the inbound group sessions linked to a room
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/traits.rs]:
```suggestion
    /// Get all the inbound group sessions for a given room.
```

File: crates/matrix-sdk-crypto/src/store/traits.rs [hunk 4]
```diff
@@ -121,6 +121,15 @@ pub trait CryptoStore: AsyncTraitDeps {
...
+    ///
+    /// # Arguments
+    /// * `room_id` - the room id of the room that the sessions are linked to.
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/traits.rs]:
```suggestion
    /// * `room_id` - The ID of the room to return sessions for.
```

File: crates/matrix-sdk-crypto/src/store/memorystore.rs [hunk 5]
```diff
@@ -443,6 +443,26 @@ impl CryptoStore for MemoryStore {
...
+    async fn get_inbound_group_sessions_by_room_id(
+        &self,
+        room_id: &RoomId,
+    ) -> Result<Vec<InboundGroupSession>> {
+        let inbounds: Vec<InboundGroupSession> = self
+            .inbound_group_sessions
+            .read()
+            .get(room_id)
+            .into_iter()
+            .flat_map(HashMap::values)
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/memorystore.rs]: Converting the `Option` into an iterator seems like a bit of an obscure way of dealing with it. Maybe you could do something with `map` or a `match` statement.

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 6]
```diff
@@ -567,6 +567,64 @@ macro_rules! cryptostore_integration_tests {
                 assert_eq!(store.inbound_group_session_counts(None).await.unwrap().total, 1);
             }
 
+            #[async_test]
+            async fn test_load_inbound_group_sessions_by_room_id() {
+                let dir = "load_inbound_group_session_by_room_id";
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: I don't know why the other tests are called `load_*`, but there doesn't seem to be any reason to propagate it.

```suggestion
            async fn test_get_inbound_group_sessions_by_room_id() {
                let dir = "get_inbound_group_session_by_room_id";
```

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 7]
```diff
@@ -567,6 +567,64 @@ macro_rules! cryptostore_integration_tests {
...
+                assert_eq!(store.get_inbound_group_sessions().await.unwrap().len(), 0);
+
+                let room_id = &room_id!("!testing:localhost");
+                let (_, session_1) = account.create_group_session_pair_with_defaults(room_id).await;
+                let (_, session_2) = account.create_group_session_pair_with_defaults(room_id).await;
+                let (_, session_3) = account.create_group_session_pair_with_defaults(room_id).await;
+                let (_, session_4) = account.create_group_session_pair_with_defaults(room_id).await;
+                let (_, session_5) = account.create_group_session_pair_with_defaults(room_id).await;
+                let (_, session_6) = account.create_group_session_pair_with_defaults(room_id).await;
+                let (_, session_7) = account.create_group_session_pair_with_defaults(room_id).await;
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: I'm not convinced we need 7 sessions to test with. How about 2?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 8]
```diff
@@ -567,6 +567,64 @@ macro_rules! cryptostore_integration_tests {
...
+
+
+                let mut sessions = vec![
+                    session_1.clone(),
+                    session_2.clone(),
+                    session_3.clone(),
+                    session_4.clone(),
+                    session_5.clone(),
+                    session_6.clone(),
+                    session_7.clone(),
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: I think the `.clone` calls are redundant?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 7 above)
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: I think we also need a session for a *different* room, to assert that is *not* returned.

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 9]
```diff
@@ -567,6 +567,64 @@ macro_rules! cryptostore_integration_tests {
...
+                    .unwrap();
+
+                for session in loaded_sessions.iter() {
+                    session.export().await;
+                }
+                assert_eq!(loaded_sessions.len(), 7);
+                assert_session_lists_eq(sessions, loaded_sessions, "room by id sessions");
+
+                assert_eq!(store.get_inbound_group_sessions().await.unwrap().len(), 7);
+                assert_eq!(store.inbound_group_session_counts(None).await.unwrap().total, 7);
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: this seems unnecessary and unrelated to the function under test?

File: crates/matrix-sdk-crypto/src/store/memorystore.rs (hunk 5 above)
Comment (multisme) [reply crates/matrix-sdk-crypto/src/store/memorystore.rs]: I changed this and use an expect that panic now.

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 9 above)
Comment (multisme) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs]: I kept it as to show the session in the other room is inside the store.

File: crates/matrix-sdk-crypto/src/store/memorystore.rs [hunk 10]
```diff
@@ -485,6 +465,26 @@ impl CryptoStore for MemoryStore {
...
 
+    async fn get_inbound_group_sessions_by_room_id(
+        &self,
+        room_id: &RoomId,
+    ) -> Result<Vec<InboundGroupSession>> {
+        let inbounds: Vec<InboundGroupSession> = self
+            .inbound_group_sessions
+            .read()
+            .get(room_id)
+            .expect("No inbound group sessions for that roomId")
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/memorystore.rs]: Panicking in rust should happen only in situations that you strongly believe should not happen. Having no sessions for a given room is a perfectly normal situation.

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 11]
```diff
@@ -604,8 +597,13 @@ macro_rules! cryptostore_integration_tests {
                 };
                 store.save_changes(changes).await.expect("Can't save group session");
 
+
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: why did you add this extra blank line?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 12]
```diff
@@ -568,29 +568,22 @@ macro_rules! cryptostore_integration_tests {
...
-                    session_4.clone(),
-                    session_5.clone(),
-                    session_6.clone(),
-                    session_7.clone(),
+                    session_1,
+                    session_2,
+                    session_3
                 ];
 
                 for session in sessions.iter_mut() {
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: can't this be simply:

```suggestion
                for session in sessions {
```

?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 13]
```diff
@@ -604,8 +597,13 @@ macro_rules! cryptostore_integration_tests {
                 };
                 store.save_changes(changes).await.expect("Can't save group session");
 
+
                 drop(store);
 
+                // Remove the last sessions since it is in an other room
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: ... so?

```suggestion
                // The last session is in a different room, so should not be returned by 
                // get_inbound_group_sessions_by_room_id. Remove it from the list.
```

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 14]
```diff
@@ -618,11 +616,12 @@ macro_rules! cryptostore_integration_tests {
                 for session in loaded_sessions.iter() {
                     session.export().await;
                 }
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: what is the purpose of this loop?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 9 above)
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs]: fair enough, but:
 * we only need one of these calls to prove that
 * it would be more logical to put it nearer the start of the test (just after you have saved them in the store)
 * please add a comment to say what it is for

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 15]
```diff
@@ -604,8 +597,13 @@ macro_rules! cryptostore_integration_tests {
                 };
                 store.save_changes(changes).await.expect("Can't save group session");
 
+
                 drop(store);
 
+                // Remove the last sessions since it is in an other room
+                sessions.pop();
+
+
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: again, why two blank lines here?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 16]
```diff
@@ -568,29 +568,22 @@ macro_rules! cryptostore_integration_tests {
             }
 
             #[async_test]
-            async fn test_load_inbound_group_sessions_by_room_id() {
-                let dir = "load_inbound_group_session_by_room_id";
+            async fn test_get_inbound_group_sessions_by_room_id() {
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs:581]: please could you add another test that checks we get an empty list back when there are no sessions in the room?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 12 above)
Comment (multisme) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs]: I need to keep the sessions to compare later with loaded _sessions. while exporting them.

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 14 above)
Comment (multisme) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs]: To export the loaded_sessions before making a comparaison with the sessions generated at the beginning of the file

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 16 above)
Comment (multisme) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs:581]: Done

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 17]
```diff
@@ -586,7 +586,7 @@ macro_rules! cryptostore_integration_tests {
                     session_3
                 ];
 
-                for session in sessions.iter_mut() {
+                for session in sessions.iter_mut(){
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: :-1:

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 18]
```diff
@@ -619,9 +621,7 @@ macro_rules! cryptostore_integration_tests {
 
                 assert_eq!(loaded_sessions.len(), 2);
                 assert_session_lists_eq(sessions, loaded_sessions, "room by id sessions");
-
-                assert_eq!(store.get_inbound_group_sessions().await.unwrap().len(), 3);
-                assert_eq!(store.inbound_group_session_counts(None).await.unwrap().total, 3);
+
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: redundant blank line here

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs [hunk 19]
```diff
@@ -574,6 +574,8 @@ macro_rules! cryptostore_integration_tests {
                 assert_eq!(store.get_inbound_group_sessions().await.unwrap().len(), 0);
 
                 let room_id = &room_id!("!testing:localhost");
+                assert_eq!(store.get_inbound_group_sessions_by_room_id(room_id).await.unwrap().len(), 0);
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/store/integration_tests.rs]: this should really have been a separate test, but I guess it's fine.

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 14 above)
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs]: Ok, but why? Export it to where? Isn't `session.export()` a no-op if you throw away the result?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 12 above)
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs]: Actually, my question should have been: what is the point of this loop? Why do we need to `export` and re-import the sessions? Can't we just remove this loop?

File: crates/matrix-sdk-crypto/src/store/integration_tests.rs (hunk 14 above)
Comment (multisme) [reply crates/matrix-sdk-crypto/src/store/integration_tests.rs]: A test was not working without but I can't reproduce it so I remove it.

**Comment by richvdh** [comment]
@multisme if you could click "re-request review" after you make changes, that would help me remember to take another look
